Handle LexisNexis giving us an error 'document' instead of an article

pmyteh opened the pull request:
Under some circumstances LexisNexis provides an error message as one of the documents in a downloaded file. The offending section of the file looks like the following:
```
<A NAME="DOC_ID_169_0"></A><!-- Hide XML section from browser
<DOC NUMBER=1>
<DOCFULL> -->
<BR><DIV CLASS="c5"><P CLASS="c6"><SPAN CLASS="c2">We are sorry but there is an error in this document and it is not possible to display it.</SPAN></P>
</DIV>
<!-- Hide XML section from browser
</DOCFULL>
</DOC> -->
```
This causes the whole file to fail; `readLexisNexisHTML()` raises an error as it can't find a copyright string.

This patch amends `LexisNexisSource()` to check for the error message and drop the failed document with a warning.

The test is currently localised only to English as I don't know what the appropriate message is in other languages. An alternative approach would be to handle the no-copyright-notice case in `readLexisNexisHTML()` (which would catch a wider variety of malformed documents) but I think then the only option would be to feed an empty `PlainTextDocument` into the corpus, which is undesirable.

Edit: There turned out to be one document in my corpus that had a malformed copyright statement even without this error, so I've added a second patch to handle this case with a warning.

nalimilan commented on tm.plugin.lexisnexis/R/readLexisNexisHTML.R:
```diff
@@ -56,7 +56,13 @@ readLexisNexisHTML <- FunctionGenerator(function(elem, language, id) {
+            copyright <- vals[[max(cr)]]
+        } else {
+            warning(sprintf("Could not parse copyright notice for article %s. This may indicate a problem with the source data, as LexisNexis copyright notices are nearly universal.\n", id))
+            copyright=NULL
```
```suggestion
            copyright <- NULL
```

nalimilan commented on tm.plugin.lexisnexis/R/readLexisNexisHTML.R:
```diff
@@ -56,7 +56,13 @@ readLexisNexisHTML <- FunctionGenerator(function(elem, language, id) {
+        if (any(cr)) {
+            copyright <- vals[[max(cr)]]
+        } else {
+            warning(sprintf("Could not parse copyright notice for article %s. This may indicate a problem with the source data, as LexisNexis copyright notices are nearly universal.\n", id))
```
Does it really indicate a problem with the source? In your case, do you have any idea what produced the problem?

pmyteh replied:
My dataset has 48 unparseable copyright notices, of which 47 are instances of an 'error' document like the one in the PR comment (at last once from six of my 23 newspapers), and one is a Washington Post article which is otherwise complete and valid but has a "Copy 2013 Washington Post ..." copyright notice rather than "Copyright 2013 Washington Post ...", so it doesn't trigger the regex.

So I suspect that most examples of this code being triggered will be non-English users hitting error message. But it's possible there are a few other malformed documents like the WaPo one.

nalimilan commented on tm.plugin.lexisnexis/R/LexisNexisSource.R:
```diff
@@ -21,6 +21,15 @@ LexisNexisSource <- function(x, encoding = "UTF-8") {
+                      content,
+                      fixed=TRUE)
+    if(any(errtexts)) {
+        warning(paste0(x, ": LexisNexis failed to provide some documents; skipping number(s) ", paste0(which(errtexts), collapse=" "), "\n", collapse=""))
```
This is enough AFAICT:
```suggestion
        warning(x, ": LexisNexis failed to provide some documents; skipping number(s) ",
                paste0(which(errtexts), collapse=", "))
```

pmyteh commented:
I don't have much experience of writing tests, but I'll try to put something together this afternoon. I can certainly provide test docs.

nalimilan commented:
> I don't have much experience of writing tests, but I'll try to put something together this afternoon. I can certainly provide test docs.

I think you can just check that the file loads correctly and contains the expected number of documents (see existing tests). The difficulty is to anonymize articles a bit to avoid copyright issues (ironically). No need to add docs.

pmyteh commented:
I've added a suitable example HTML file showing both kinds of problem, and added a few lines to `import.R` to load it in and then check that the correct results are present (i.e. one document dropped, the other read minus the malformed copyright notice).

I've also merged in the changes from #6 onto this branch, so the CI should work again.

nalimilan commented:
Thanks! I've submitted version 1.4.1 to CRAN.
